src/gt4py/next/ffront/stages.py

```diff
@@ -15,29 +15,134 @@
 from __future__ import annotations
 
 import dataclasses
+import inspect
 import types
-from typing import Any, Optional
+from typing import Any, Generic, Optional, TypeVar
 
 from gt4py.next import common
-from gt4py.next.ffront import program_ast as past
+from gt4py.next.ffront import field_operator_ast as foast, program_ast as past
+from gt4py.next.type_system import type_specifications as ts
+
+
+OperatorNodeT = TypeVar("OperatorNodeT", bound=foast.LocatedNode)
+
+
+class ContentHashableMixin:
+    """
+    Allows deriving dataclasses to modify what goes into the content hash per-field.
+
+    Warning: Using this will modify how the class gets pickled. If unpickling is desired,
```

Comment:
@egparedes Can you have a look at this part? I am not sure I understand its implication and it seems like a second pair of eyes make sense.

Reply:
IIUC, the `Mixin` classes and the functions have been added here to support hashing the Stage classes with `eve.content_hash` but I don't really see why this is needed. I mean, if these classes need to be hashable and there is an unique way to implement the hash that makes sense in all use cases, then we could just write a custom `__hash__` instead of hacking around with `__setstate__` and `__getstate__` methods (which in addition of breaking the pickleablity of the instances, it'd only work for the `content_hash` function).

An even better and more general solution, in my opinion, would be to define custom hash functions as required for different purposes as `singledispatch` free functions, and then register specific implementations for the classes that need to be hashable in that specific way.

Reply:
I think the singledispatch function is probably the solution.

Reply:
My solution is up for review.